rust/src/dns/dns.rs

```diff
@@ -0,0 +1,907 @@
+        match dns_parse_request(input) {
+            nom::IResult::Done(_, request) => {
+                return self.handle_request(request);
+            },
+            _ => {
+                self.events.push(DNSEvents::MalformedData);
+                return 0;
+            }
+        }
+    }
+
+    /// Parse and handle a DNS response.
+    pub fn parse_response(&mut self, input: &[u8]) -> u64 {
+        match dns_parse_response(input) {
+            nom::IResult::Done(_, response) => {
+                return self.handle_response(response);
+            },
+            _ => {
+                self.events.push(DNSEvents::MalformedData);
+                return 0;
+            }
+        }
+    }
+
+}
+
+/// Expose DNSState::new.
+export_state_new!(rs_dns_state_new, DNSState);
+
+/// Expose a function to free DNSState.
+export_state_free!(rs_dns_state_free, DNSState);
+
+#[no_mangle]
+pub extern fn rs_dns_state_tx_free(this: &mut DNSState, tx_id: libc::uint64_t) {
+   this.tx_free(tx_id);
+}
+
+#[no_mangle]
+pub extern fn rs_dns_state_tx_get(this: &mut DNSState, tx_id: libc::uint64_t)
+                                  -> *mut DNSTransaction {
+    match this.tx_get(tx_id) {
+        Some(tx) => {
+            return unsafe{transmute(tx)};
+        },
+        _ => {
+            return ptr::null_mut();
+        }
+    }
+}
+
+#[no_mangle]
+pub extern fn rs_dns_state_get_next_event(this: &mut DNSState) -> u32 {
+    match this.events.pop() {
+        Some(ev) => {
+            return ev as u32;
+        },
+        _ => {
+            return 0;
+        }
+    }
+}
+
+#[no_mangle]
+pub extern fn rs_dns_state_get_tx_count(this: &mut DNSState) -> u64 {
+    return this.tx_id;
+}
+
+#[no_mangle]
+pub extern fn rs_dns_state_parse_request(state: &mut DNSState,
+                                         input: *const libc::uint8_t,
+                                         len: libc::uint32_t) -> libc::uint64_t
+{
+    let buf = unsafe{std::slice::from_raw_parts(input, len as usize)};
+    let tx_id = state.parse_request(buf);
+    return tx_id as libc::uint64_t;
+}
+
+#[no_mangle]
+pub extern fn rs_dns_state_parse_response(state: &mut DNSState,
+                                          input: *const libc::uint8_t,
+                                          len: libc::uint32_t) -> libc::uint64_t
+{
+    let buf = unsafe{std::slice::from_raw_parts(input, len as usize)};
+    let tx_id = state.parse_response(buf);
+    return tx_id as libc::uint64_t;
+}
+
+impl DNSTransaction {
+
+    pub fn new() -> DNSTransaction {
+        return DNSTransaction{
+            id: 0,
+            request: None,
+            response: None,
+            replied: false,
+        }
+    }
+
+}
+
+#[no_mangle]
+pub extern fn rs_dns_tx_get_alstate_progress(tx: &mut DNSTransaction,
+                                             direction: libc::uint8_t)
+                                             -> libc::uint8_t {
+    if direction & core::STREAM_TOCLIENT > 0 {
```

Comment:
I would like to make this direction simpler. Like just passing 0 or 1 from the C code.